internal/pkg/deploy/pipeline.go

```diff
@@ -442,41 +445,143 @@ func (s *GitHubV1Source) Owner() (string, error) {
-func (s *PipelineStage) WorkloadTemplateConfigurationPath(wlName string) string {
-	return fmt.Sprintf(WorkloadCfnTemplateConfigurationNameFormat,
-		wlName, s.Name,
-	)
+// Name returns the stage's name.
+func (stg *PipelineStage) Name() string {
+	return stg.associatedEnvironment.Name
 }
 
-// AssociatedEnvironment defines the necessary information a pipeline stage
-// needs for an Config Environment.
-type AssociatedEnvironment struct {
-	// Name of the environment, must be unique within an application.
-	// This is also the name of the pipeline stage.
-	Name string
+// RequiresApproval returns true if the stage requires a manual approval before running deployments.
+func (stg *PipelineStage) RequiresApproval() bool {
+	return stg.requiresApproval
+}
 
-	// The region this environment is stored in.
-	Region string
+func (stg *PipelineStage) Region() string {
+	return stg.associatedEnvironment.Region
+}
 
-	// AccountID of the account this environment is stored in.
-	AccountID string
+// ExecRoleARN returns the IAM role assumed by CloudFormation to create or update resources defined in a template.
+func (stg *PipelineStage) ExecRoleARN() string {
+	return stg.execRoleARN
+}
+
+// EnvManagerRoleARN returns the IAM role used to create or update CloudFormation stacks in an environment.
+func (stg *PipelineStage) EnvManagerRoleARN() string {
+	return stg.envManagerRoleARN
+}
+
+// TestCommands returns commands to run in CodeBuild once deployments finish.
+func (stg *PipelineStage) TestCommands() []string {
+	return stg.testCommands
+}
+
+// TestCommandsOrder returns the order in which the TestCommands action should run in the stage.
+func (stg *PipelineStage) TestCommandsOrder() int {
+	order := 0
+	for _, deployment := range stg.Deployments() {
+		if cur := deployment.RunOrder(); cur > order {
+			order = cur
+		}
+	}
+	return order + 1
+}
+
+// Deployments returns a list of deploy actions for the pipeline.
+func (stg *PipelineStage) Deployments() []WorkloadDeployAction {
+	var actions []WorkloadDeployAction
+	for _, workload := range stg.localWorkloads {
+		actions = append(actions, WorkloadDeployAction{
+			name:             workload,
+			envName:          stg.associatedEnvironment.Name,
+			appName:          stg.AppName,
+			requiresApproval: stg.requiresApproval,
+		})
+	}
+	return actions
+}
+
+// WorkloadDeployAction represents a CodePipeline action of category "Deploy" for a workload stack.
+type WorkloadDeployAction struct {
+	name    string // workload name.
+	envName string
+	appName string
+
+	requiresApproval bool
+}
+
+// Name returns the name of the CodePipeline deploy action for a workload.
+func (a *WorkloadDeployAction) Name() string {
+	return fmt.Sprintf("CreateOrUpdate-%s-%s", a.name, a.envName)
+}
+
+// StackName returns the name of the workload stack to create or update.
+func (a *WorkloadDeployAction) StackName() string {
+	return fmt.Sprintf("%s-%s-%s", a.appName, a.envName, a.name)
```

Comment:
What do you think of returning `stack.NameForService(a.appName, a.envName, a.name)`?

Reply:
lol I tried this too but there is a cyclic dependency :rip: